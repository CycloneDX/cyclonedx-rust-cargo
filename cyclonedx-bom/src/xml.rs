use crate::errors::XmlWriteError;
use std::io::Write;
use xml::writer::EventWriter;

pub(crate) trait ToXmlDocument {
    fn write_xml_document<W: Write>(
        &self,
        writer: &mut EventWriter<W>,
    ) -> Result<(), XmlWriteError>;
}

pub(crate) trait ToXml {
    fn write_xml_element<W: Write>(&self, writer: &mut EventWriter<W>)
        -> Result<(), XmlWriteError>;

    fn will_write(&self) -> bool {
        true
    }
}

// TODO: is there a name for this pattern
pub(crate) trait ToInnerXml {
    fn write_xml_named_element<W: Write>(
        &self,
        writer: &mut EventWriter<W>,
        tag: &str,
    ) -> Result<(), XmlWriteError>;

    fn will_write(&self) -> bool {
        true
    }
}

pub(crate) fn to_xml_write_error(
    element: impl AsRef<str>,
) -> impl FnOnce(xml::writer::Error) -> XmlWriteError {
    let element = element.as_ref().to_owned();
    |error| XmlWriteError::XmlElementWriteError { error, element }
}

#[cfg(test)]
pub(crate) mod test {
    use xml::EmitterConfig;

    use super::*;

    fn emitter_config() -> EmitterConfig {
        EmitterConfig::default().perform_indent(true)
    }

    pub(crate) fn write_element_to_string<X: ToXml>(element: X) -> String {
        let mut output = Vec::new();
        let mut event_writer = EventWriter::new_with_config(&mut output, emitter_config());
        element
            .write_xml_element(&mut event_writer)
            .expect("Should have written the element");
        String::from_utf8_lossy(&output).to_string()
    }

    pub(crate) fn write_named_element_to_string<X: ToInnerXml>(element: X, tag: &str) -> String {
        let mut output = Vec::new();
        let mut event_writer = EventWriter::new_with_config(&mut output, emitter_config());
        element
            .write_xml_named_element(&mut event_writer, tag)
            .expect("Should have written the element");
        String::from_utf8_lossy(&output).to_string()
    }
}
