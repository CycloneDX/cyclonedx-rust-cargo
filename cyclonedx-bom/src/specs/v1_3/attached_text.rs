/*
 * This file is part of CycloneDX Rust Cargo.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

use crate::{
    errors::XmlWriteError, external_models::normalized_string::NormalizedString, xml::ToInnerXml,
};
use crate::{models, xml::to_xml_write_error};
use serde::{Deserialize, Serialize};
use xml::writer::{EventWriter, XmlEvent};

#[derive(Debug, Deserialize, Serialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub(crate) struct AttachedText {
    #[serde(skip_serializing_if = "Option::is_none")]
    content_type: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    encoding: Option<String>,
    content: String,
}

impl From<models::attached_text::AttachedText> for AttachedText {
    fn from(other: models::attached_text::AttachedText) -> Self {
        Self {
            content_type: other.content_type.map(|n| n.0),
            encoding: other.encoding.map(|e| e.to_string()),
            content: other.content,
        }
    }
}

impl From<AttachedText> for models::attached_text::AttachedText {
    fn from(other: AttachedText) -> Self {
        Self {
            content_type: other.content_type.map(NormalizedString::new_unchecked),
            encoding: other
                .encoding
                .map(models::attached_text::Encoding::new_unchecked),
            content: other.content,
        }
    }
}

impl ToInnerXml for AttachedText {
    fn write_xml_named_element<W: std::io::Write>(
        &self,
        writer: &mut EventWriter<W>,
        tag: &str,
    ) -> Result<(), XmlWriteError> {
        let mut attached_text_tag = XmlEvent::start_element(tag);

        if let Some(content_type) = &self.content_type {
            attached_text_tag = attached_text_tag.attr("content-type", content_type);
        }

        if let Some(encoding) = &self.encoding {
            attached_text_tag = attached_text_tag.attr("encoding", encoding);
        }
        writer
            .write(attached_text_tag)
            .map_err(to_xml_write_error(tag))?;

        writer
            .write(XmlEvent::characters(&self.content))
            .map_err(to_xml_write_error(tag))?;
        writer
            .write(XmlEvent::end_element())
            .map_err(to_xml_write_error(tag))?;

        Ok(())
    }
}

#[cfg(test)]
pub(crate) mod test {
    use super::*;
    use crate::xml::test::write_named_element_to_string;

    pub(crate) fn example_attached_text() -> AttachedText {
        AttachedText {
            content_type: Some("content type".to_string()),
            encoding: Some("encoding".to_string()),
            content: "content".to_string(),
        }
    }

    pub(crate) fn corresponding_attached_text() -> models::attached_text::AttachedText {
        models::attached_text::AttachedText {
            content_type: Some(NormalizedString::new_unchecked("content type".to_string())),
            encoding: Some(models::attached_text::Encoding::UnknownEncoding(
                "encoding".to_string(),
            )),
            content: "content".to_string(),
        }
    }

    #[test]
    fn it_should_write_xml_full() {
        let xml_output = write_named_element_to_string(example_attached_text(), "text");
        insta::assert_snapshot!(xml_output);
    }

    #[test]
    fn it_should_write_xml_no_attributes() {
        let xml_output = write_named_element_to_string(
            AttachedText {
                content_type: None,
                encoding: None,
                content: "content".to_string(),
            },
            "text",
        );
        insta::assert_snapshot!(xml_output);
    }
}
