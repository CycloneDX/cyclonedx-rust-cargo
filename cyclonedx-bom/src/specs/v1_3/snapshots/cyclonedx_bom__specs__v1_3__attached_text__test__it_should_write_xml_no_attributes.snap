---
source: cyclonedx-bom/src/specs/v1_3/attached_text.rs
assertion_line: 132
expression: xml_output

---
<?xml version="1.0" encoding="utf-8"?>
<text>content</text>
